Automatically testing a changed process/subworkflow and its dependencies

simonleandergrimm opened the pull request:
0th draft. See thread for context: https://twist.com/a/197793/inbox/t/6808534/

harmonbhasin reviewed (approved):
LGTM!

harmonbhasin reviewed (changes requested):
Changes this to 'Request changes' for the CHANGELOG update.

simonleandergrimm commented:
@harmonbhasin Can you assign to Will once you're happy with this?

simonleandergrimm commented:
I just noticed that we don't check the workflow runs anymore as part of the Github-side tests. Will add those back in.

simonleandergrimm commented:
Just noting that this is in my backlog, will get to it.

simonleandergrimm commented:
Just make sure to assign it to me when you're ready for me to look at it again.
Yeah, will do!

simonleandergrimm commented:
@harmonbhasin Updated the script and reassigned it to you. Also, do you think the script should go under bin/?

harmonbhasin commented:
@simonleandergrimm yes i think putting it under `bin/` would be good.

harmonbhasin reviewed (dismissed):
Given that you've confirmed that this script works as intended locally, this looks good to me.

simonleandergrimm commented:
@harmonbhasin Can you run it locally too?

harmonbhasin commented:
@simonleandergrimm yes let me try this now

harmonbhasin commented:
@simonleandergrimm this worked well for me, just move the script into the bin folder, and you're good to merge.

willbradshaw commented:
@harmonbhasin @simonleandergrimm noting that I want to review this (and all PRs) before merging into dev. Let me know when it's ready for me to do that.

harmonbhasin commented:
@willbradshaw sounds good, this is ready for you to look at.

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,160 @@
+    direct_tests = find_dependency("tests/", component)
+    tests_to_execute.update(direct_tests)
+
+    print("=" * 80)
```
This should probably be 72: https://peps.python.org/pep-0008/#maximum-line-length

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,160 @@
+    for test in direct_tests:
+        print(f"   • {test}")
+    print()
+    # print("\n")
```
Delete or uncomment

simonleandergrimm replied:
Done

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,160 @@
+    return component, verbose
+
+
+def main():
```
This function needs comments

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,160 @@
+
+    test_results = {}
+    for test in sorted(tests_to_execute):
+        args = ["nf-test", "test", test]
```
I'm worried that running a separate nf-test command for every test will be substantially slower than a single command (or a small number of commands) passed multiple tests. I'm not confident about that but I think it's worth checking before we commit to this approach.

OTOH this format allows us to add in parallelisation later, which is great.

willbradshaw replied:
> OTOH this format allows us to add in parallelisation later, which is great.

Speaking of which, @harmonbhasin we should integrate this with the functionality from your https://github.com/naobservatory/mgs-workflow/blob/master/bin/run_parallel_test.sh. No point having two competing test handlers, it will just confuse people.

simonleandergrimm replied:
@harmonbhasin can you take that on or shall I.

simonleandergrimm replied:
Having looked into this, I'm unsure if nf-test --shard is the right approach here? We could just use Python's multiprocessing

simonleandergrimm replied:
Given that we are passings tests individually, not as dirs.

simonleandergrimm replied:
@harmonbhasin Might be most efficient to briefly chat about this on Tuesday.

willbradshaw replied:
Yeah, if you're calling each test file (I believe you're passing test files, not individual tests?) separately, I don't see much advantage of nf-test --shard over another parallelisation approach.

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,160 @@
+            args.append("--verbose")
+        print(f"\nRunning \"{' '.join(args)}\"")
+        print("-" * 80)
+        result = subprocess.run(args)
```
@harmonbhasin and I went back and forth a fair bit about how test handlers should deal with output from tests (pass to stderr? write to a file? etc) for run_parallel_test.sh, I recommend taking the same approach here.

willbradshaw replied:
Also relevant: `execute_single` method from https://github.com/naobservatory/mgs-orchestrator/blob/main/orchestrator/pipeline_executor.py

willbradshaw replied:
(Though in this case I probably recommend writing to a file rather than passing through directly, or maybe capturing and then writing the output for failed tests at the end)

simonleandergrimm replied:
If we implement parallelisation, putting the stdout in log, with stderr printed to terminal upon completion (and keeping the summary I give with which tests worked vs which didn't).

willbradshaw replied:
Even without parallelisation, I think you need to think about how best to capture, aggregate and return the results of all these subprocess calls.

simonleandergrimm replied:
FWIW, I already do that (lmk if you don't like how it's currently looking).

willbradshaw replied:
This is now moot since we're just making one big call to nf-test, so I'll resolve this for now, but we'll need to come back to this when we introduce parallelisation.

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,160 @@
+        print(f"   • {test}")
+    print()
+
+    workflow_tests = set()
```
Note that workflows can take modules as dependencies, not just subworkflows, so need to check those as well

simonleandergrimm replied:
Added

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,160 @@
+        "-c",
+        "--component",
+        required=True,
+        help="Name of component (process or subworkflow) to test",
```
Is this intended to be a name or a filepath? It says name, but in some functions above it gets treated like a filepath.

I'd recommend a name, e.g. `EXTRACT_VIRAL_READS`, rather than a path, e.g. `subworkflows/local/extractViralReads/main.nf`

If you go with a name you can enforce it, e.g. by requiring it to only include alphanumerics and underscores.

simonleandergrimm replied:
Good point yeah, it should be something like EXTRACT_VIRAL_READS.

simonleandergrimm replied:
Added something to this effect:
```
[ec2-user@ip-172-31-40-150 mgs-workflow]$ bin/test_component_dependencies.py -c FASTP/
Error: Input component 'FASTP/' must be a process or subworkflow name, not the file path that contains said process/subworkflow. E.g., 'FASTP' not 'modules/local/fastp/main.nf'
```

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,160 @@
+    component = component.replace("/main.nf", "")
+
+    grep_results = subprocess.run(
+        ["grep", "-r", component, directory], capture_output=True, text=True
```
Just grepping like this will fail in cases where one component is a substring of another, e.g. RUN vs RUN_VALIDATION.

willbradshaw replied:
I'd suggest only looking in specific lines, e.g. `include` statements and setup-block `run` statements.

Oh, and nf-test `workflow` / `process` declarations.

simonleandergrimm replied:
I fixed this with word boundaries, let me know if this still addresses your concern.

willbradshaw replied:
I agree word boundaries solve the narrow problem I was pointing to here, but there are other ways this check is overbroad, e.g. what if a process is mentioned in a comment in another process? That would be enough to trigger this even if the process isn't actually called anywhere in the file.

simonleandergrimm replied:
Good point. I added further logic that should stop that from happening.

simonleandergrimm replied:
Though I didn't know what you meant by `run` statements.

willbradshaw replied:
By `run` statements, I mean statements like the following within `setup` blocks in tests:

```
setup {
    run("LOAD_DOWNSTREAM_DATA") {
        // ...
    }
}
```

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,160 @@
+    with open(workflow_path, "r") as f:
+        workflow_content = f.read()
+
+    if subworkflow_dir in workflow_content:
```
Here also you're vulnerable to substring matches

simonleandergrimm replied:
See my longer comment below.

simonleandergrimm commented:
@willbradshaw for your concerns regarding substring matching, instead of what you suggested I implement grepping/regex that uses word boundaries.

E.g.: with the faulty subworkflow `subworkflows/local/profi/main.nf`, the original code would have returned True:

```{test-script.py}
#! /usr/bin/env python3
import re

workflow = "run.nf"
subworkflow_path = "subworkflows/local/profi/main.nf"
workflow_path = f"workflows/{workflow}"
subworkflow_dir = subworkflow_path.replace("/main.nf", "")

with open(workflow_path, "r") as f:
    workflow_content = f.read()

# Use word boundaries to preempt substring matches
if subworkflow_dir in workflow_content:
    print("True")
else:
    print("False")
```
```
[ec2-user@ip-172-31-40-150 mgs-workflow]$ ./test-script.py 
True
```

However, with word boundaries, this is fixed:

```{test-script-boundary.py}
#! /usr/bin/env python3
import re

workflow = "run.nf"
subworkflow_path = "subworkflows/local/profi/main.nf"
workflow_path = f"workflows/{workflow}"
subworkflow_dir = subworkflow_path.replace("/main.nf", "")

with open(workflow_path, "r") as f:
    workflow_content = f.read()

# Use word boundaries to preempt substring matches
if re.search(r'\b' + re.escape(subworkflow_dir) + r'\b', workflow_content):
    print("True")
else:
    print("False")
```

```
[ec2-user@ip-172-31-40-150 mgs-workflow]$ ./test-script-boundary.py 
False
```


I did the same for grepping the component names.

Now when using the subcomponent name "FAST" I get no tests returned:

```
[ec2-user@ip-172-31-40-150 mgs-workflow]$ bin/test_component_dependencies.py -c FAST
{.....}

----------------------------------------
Total:  0 tests
Passed: 0 tests
Failed: 0 tests
----------------------------------------
```

Whereas previously we did:

```
[ec2-user@ip-172-31-40-150 mgs-workflow]$ bin/test_component_dependencies.py -c FAST
{....}
Identified 38 tests to execute. Running...
{....}
```

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,177 @@
+
+    # Validate component name - only allow alphanumerics and underscores
+    if not re.match(r'^[a-zA-Z0-9_]+$', component):
+        print(f"Error: Input component '{component}' must be a process or subworkflow name, not the file path that contains said process/subworkflow. E.g., 'FASTP' not 'modules/local/fastp/main.nf'")
```
I think this is good to have but the message is a bit too specific for the condition that triggers it. E.g. if someone specified "ABC+" they'd be told "you can't use a path here", but they didn't.

Also, you should either use an `assert` or a `raise` here, not a print + exit.

simonleandergrimm replied:
Made both fixes.

willbradshaw commented:
My main concern about not implementing parallelisation now is this earlier comment:

> I'm worried that running a separate nf-test command for every test will be substantially slower than a single command (or a small number of commands) passed multiple tests. I'm not confident about that but I think it's worth checking before we commit to this approach.

simonleandergrimm commented:
> My main concern about not implementing parallelisation now is this earlier comment:
> 
> I'm worried that running a separate nf-test command for every test will be substantially slower than a single command (or a small number of commands) passed multiple tests. I'm not confident about that but I think it's worth checking before we commit to this approach.
> 
> If that turns out not to be an issue I agree we don't lose much by just merging this in serial mode and then adding parallel processing later.

Running 6 tests with launching nf-test individually added 23 seconds, so around 4 seconds per test. Seems fine?
```
Approach 1 (All at once):
  - Total time: 0:02:05.100760
  - Status: SUCCESS

Approach 2 (Individual runs):
  - Total time: 0:02:28.822050
  - Successful tests: 6/6

Time difference:
  - Running all tests at once was 0:00:23.721289 faster
  - Improvement: 18.96%
```

willbradshaw commented:
That doesn't seem fine to me! A ~20% increase for just six tests, with the expectation that it will get worse with more tests run, is pretty bad. We care quite a bit about performance for testing.

The alternative, which is less amenable to parallel processing but better for serial processing, is to just collate a bunch of paths in a list and then pass them all to nf-test in a single call. That seems like it should be pretty easy to implement.

simonleandergrimm commented:
Alright, I've now done the thing of collating paths into a list and passing that to nf-test. The performance for that is fine:
```
================================================================================
SUMMARY
================================================================================
Approach 1 (All at once with directory):
  - Total time: 0:01:49.218480
  - Status: SUCCESS

Approach 2 (All at once with explicit test files):
  - Total time: 0:01:48.840106
  - Status: SUCCESS

Time difference:
  - Running with explicit files was 0:00:00.378374 faster
  - Improvement: 0.35%
 ```

simonleandergrimm commented:
@willbradshaw This is ready for re-review. Additionally, it'd be good if you ran the script on your side to spot any final improvements you'd like to be made.

willbradshaw commented on bin/test_component_dependencies.py:
```diff
@@ -0,0 +1,166 @@
+    print("=" * 72)
+    print(f"Identified {len(tests_to_execute)} tests to execute. Running...")
+
+    test_results = {}
```
I think this now isn't doing anything?

willbradshaw replied:
@simonleandergrimm ping on this comment

willbradshaw commented:
OK, I tried running this on my own instance:

```
ec2-user@ip-172-31-75-190:~/mgs-workflow/repo$ ./bin/test_component_dependencies.py -c BBMERGE
========================================================================
Found 1 direct tests:
========================================================================
   • tests/modules/local/bbmerge/main.nf.test

========================================================================
Found 2 dependent subworkflow tests:
========================================================================
   • tests/subworkflows/local/taxonomy/main.nf.test
   • tests/modules/local/processKrakenViral/main.nf.test

========================================================================
Found 0 dependent workflow tests:
========================================================================

========================================================================
Identified 3 tests to execute. Running...
// Etc
```

This fundamentally works, but I have some notes:

1. I don't know why it's running `tests/modules/local/processKrakenViral/main.nf.test`; is it because it calls the taxonomy subworkflow in the setup method? And why is it calling it a subworkflow test when it's a module? Probably some clearer logging would be helpful here.
2. I think this should run the workflow that depends on the dependent subworkflows (in this case, RUN) to check that it still works in context.

simonleandergrimm commented:
> I don't know why it's running tests/modules/local/processKrakenViral/main.nf.test; is it because it calls the taxonomy subworkflow in the setup method? And why is it calling it a subworkflow test when it's a module? Probably some clearer logging would be helpful here.

Thanks for testing! This is because the logging comment was wrong: We are executing tests that _use_ that subworkflow, which can include tests of modules.

> I think this should run the workflow that depends on the dependent subworkflows (in this case, RUN) to check that it still works in context.

Yup, I agree, this misses that some subworkflows that are affected are used by another susbworkflow, which gets missed by `workflow_uses_subworkflow()`.

I added code that fixes this. Most important logic:

```
...
    # Find subworkflows that directly use the component
    dependent_subworkflows = set(find_dependency("subworkflows/", component))

    # Find transitive dependencies - subworkflows that use other dependent subworkflows
    transitive_dependencies = set()
    for subworkflow in os.listdir("subworkflows/local/"):
        for dependent_subworkflow in dependent_subworkflows:
            if subworkflow_uses_subworkflow(subworkflow, dependent_subworkflow):
                transitive_dependencies.add(subworkflow)

    # Combine all dependent subworkflows
    dependent_subworkflows.update(transitive_dependencies)

    # Identifying tests that use the affected subworkflows
    subworkflow_tests = set()
    for dependent_subworkflow in dependent_subworkflows:
        tests = find_dependency("tests/subworkflows/", dependent_subworkflow)
        subworkflow_tests.update(tests)
```

willbradshaw commented:
Hmm, now I'm getting a weird error when I try to run the script. It works at [3fe9878](https://github.com/naobservatory/mgs-workflow/pull/232/commits/3fe98784c04fd77e8cf7632a6e36bd9afe6bddb4), but breaks at [daddb29](https://github.com/naobservatory/mgs-workflow/pull/232/commits/daddb29f3a3e488d6ef696d75e91f430d2c6baff) or later.

willbradshaw commented:
OK I figured out the problem on my end and fixed it, now it works, here's another example output:

```
ec2-user@ip-172-31-75-190:~/mgs-workflow/repo$ ./bin/test_component_dependencies.py -c SUBSET_FASTQ
========================================================================
Found 2 direct tests:
========================================================================
   • tests/modules/local/.nf-test.log
   • tests/modules/local/subsetFastq/main.nf.test

========================================================================
Found 1 tests for affected subworkflows:
========================================================================
   • tests/subworkflows/local/blastViral/main.nf.test

========================================================================
Found 2 tests for affected workflows:
========================================================================
   • tests/workflows/run_validation.nf.test
   • tests/workflows/run.nf.test

========================================================================
Identified 5 test files to execute. Running...

Running test files:
   • tests/modules/local/.nf-test.log
   • tests/modules/local/subsetFastq/main.nf.test
   • tests/subworkflows/local/blastViral/main.nf.test
   • tests/workflows/run.nf.test
   • tests/workflows/run_validation.nf.test
------------------------------------------------------------------------
...
```

Looking very nice! It shouldn't be pulling in `.nf-test` files, though.

simonleandergrimm commented:
Added logic to drop those files:

```
    file_paths = set()
    for line in grep_results.stdout.splitlines():
        if any(trigger in line for trigger in relevant_triggers):
                file = line.split(":", 1)[0]
                if "nf-test" in file:
                    continue
                else:
                    file_paths.add(file)
```

willbradshaw commented:
OK great, this looks good to me now! @simonleandergrimm can you update the CHANGELOG? Then we should be good to merge.

simonleandergrimm commented:
CHANGELOG is updated.

willbradshaw commented:
@simonleandergrimm now you need to merge in dev and fix the conflicts.

simonleandergrimm commented:
Done!

willbradshaw commented:
@simonleandergrimm assuming the tests all pass, you can merge this!

willbradshaw commented:
Nice work, I'm excited to have this functionality.
